File: src/com/cattailsw/nanidroid/ViewServerLifecycle.java
package com.cattailsw.nanidroid;

import android.app.Activity;
import android.os.Build;

import com.android.debug.hv.ViewServer;

/** Keeps the legacy debug ViewServer outside modern Android activity lifecycles. */
final class ViewServerLifecycle {
    interface Backend {
        void addWindow(Activity activity);

        void setFocusedWindow(Activity activity);

        void removeWindow(Activity activity);
    }

    private static final Backend LEGACY_BACKEND = new Backend() {
        @Override
        public void addWindow(Activity activity) {
            ViewServer.get(activity).addWindow(activity);
        }

        @Override
        public void setFocusedWindow(Activity activity) {
            ViewServer.get(activity).setFocusedWindow(activity);
        }

        @Override
        public void removeWindow(Activity activity) {
            ViewServer.get(activity).removeWindow(activity);
        }
    };

    private ViewServerLifecycle() {
    }

    static void onActivityCreated(Activity activity) {
        onActivityCreated(Build.VERSION.SDK_INT, activity, LEGACY_BACKEND);
    }

    static void onActivityResumed(Activity activity) {
        onActivityResumed(Build.VERSION.SDK_INT, activity, LEGACY_BACKEND);
    }

    static void onActivityDestroyed(Activity activity) {
        onActivityDestroyed(Build.VERSION.SDK_INT, activity, LEGACY_BACKEND);
    }

    static void onActivityCreated(int sdkInt, Activity activity, Backend backend) {
        if (usesLegacyViewServer(sdkInt)) {
            backend.addWindow(activity);
        }
    }

    static void onActivityResumed(int sdkInt, Activity activity, Backend backend) {
        if (usesLegacyViewServer(sdkInt)) {
            backend.setFocusedWindow(activity);
        }
    }

    static void onActivityDestroyed(int sdkInt, Activity activity, Backend backend) {
        if (usesLegacyViewServer(sdkInt)) {
            backend.removeWindow(activity);
        }
    }

    private static boolean usesLegacyViewServer(int sdkInt) {
        return sdkInt < Build.VERSION_CODES.HONEYCOMB;
    }
}

File: src/com/cattailsw/nanidroid/ViewServerLifecycle.kt
package com.cattailsw.nanidroid

import android.app.Activity
import android.os.Build
import com.android.debug.hv.ViewServer

/** Keeps the legacy debug ViewServer outside modern Android activity lifecycles. */
object ViewServerLifecycle {
    interface Backend {
        fun addWindow(activity: Activity?)

        fun setFocusedWindow(activity: Activity?)

        fun removeWindow(activity: Activity?)
    }

    private val legacyBackend = object : Backend {
        override fun addWindow(activity: Activity?) {
            ViewServer.get(activity!!).addWindow(activity)
        }

        override fun setFocusedWindow(activity: Activity?) {
            ViewServer.get(activity!!).setFocusedWindow(activity)
        }

        override fun removeWindow(activity: Activity?) {
            ViewServer.get(activity!!).removeWindow(activity)
        }
    }

    @JvmStatic
    fun onActivityCreated(activity: Activity) {
        onActivityCreated(Build.VERSION.SDK_INT, activity, legacyBackend)
    }

    @JvmStatic
    fun onActivityResumed(activity: Activity) {
        onActivityResumed(Build.VERSION.SDK_INT, activity, legacyBackend)
    }

    @JvmStatic
    fun onActivityDestroyed(activity: Activity) {
        onActivityDestroyed(Build.VERSION.SDK_INT, activity, legacyBackend)
    }

    @JvmStatic
    fun onActivityCreated(sdkInt: Int, activity: Activity?, backend: Backend) {
        if (usesLegacyViewServer(sdkInt)) backend.addWindow(activity)
    }

    @JvmStatic
    fun onActivityResumed(sdkInt: Int, activity: Activity?, backend: Backend) {
        if (usesLegacyViewServer(sdkInt)) backend.setFocusedWindow(activity)
    }

    @JvmStatic
    fun onActivityDestroyed(sdkInt: Int, activity: Activity?, backend: Backend) {
        if (usesLegacyViewServer(sdkInt)) backend.removeWindow(activity)
    }

    private fun usesLegacyViewServer(sdkInt: Int): Boolean =
        sdkInt < Build.VERSION_CODES.HONEYCOMB
}
